pysat/instruments/sw_f107.py

```diff
@@ -357,27 +440,151 @@ def download(date_array, tag, sat_id, data_path, user=None, password=None):
-        f107 = []
-        f107_times = []
-        for line in raw_f107:
-            for i in np.arange(5):
-                f107_times.append(pysat.datetime.strptime(line[0:7], '%d%b%y'))
-                f107.append(int(line[8:11]))
-        
+        ap_times, ap = parse_45day_block(raw_ap)
+
+        # parse the F10.7 data
+        f107_times, f107 = parse_45day_block(raw_f107)
+
         # collect into DataFrame
         data = pds.DataFrame(f107, index=f107_times, columns=['f107'])
         data['ap'] = ap
         # write out as a file
         data.to_csv(os.path.join(data_path, 'f107_45day_' +
                                  date.strftime('%Y-%m-%d') + '.txt'))
 
-    return        
+    return
+
+def parse_45day_block(block_lines):
+    """ Parse the data blocks used in the 45-day Ap and F10.7 Flux Forecast file
+
+    Parameters
+    ----------
+    block_lines : (list)
+        List of lines containing data in this data block
+
+    Returns
+    -------
+    dates : (list)
+        List of dates for each date/data pair in this block
+    values : (list)
+        List of values for each date/data pair in this block
+
+    """
+
+    # Initialize the output
+    dates = list()
+    values = list()
+    
+    # Cycle through each line in this block
+    for line in block_lines:
+        # Split the line on whitespace
+        split_line = line.split()
+
+        # Format the dates
+        dates.extend([pysat.datetime.strptime(tt, "%d%b%y")
+                      for tt in split_line[::2]])
+
+        # Format the data values
+        values.extend([int(vv) for vv in split_line[1::2]])
+
+    return dates, values
+
+def parse_daily_solar_data(data_lines):
+    """ Parse the data in the SWPC daily solar index file
+
+    Parameters
+    ----------
+    data_lines : (list)
+        List of lines containing data
+
+    Returns
+    -------
+    dates : (list)
+        List of dates for each date/data pair in this block
+    values : (dict)
+        Dict of lists of values, where each key is the value name
+
+    """
+
+    # Initialize the output
+    dates = list()
+    val_keys = ['f107', 'ssn', 'ss_area', 'new_reg', 'smf', 'goes_bgd_flux',
+                'c_flare', 'm_flare', 'x_flare', 'o1_flare', 'o2_flare',
+                'o3_flare']
+    values = {kk: list() for kk in val_keys}
+    
+    # Cycle through each line in this file
+    for line in data_lines:
+        # Split the line on whitespace
+        split_line = line.split()
+
+        # Format the date
+        dates.append(pysat.datetime.strptime("".join(split_line[0:3]),
+                                                     "%Y%m%d"))
+
+        # Format the data values
+        for i,kk in enumerate(val_keys):
+            val = split_line[i + 3]
+            if kk != 'goes_bgd_flux':
+                val = int(val)
+            values[kk].append(val)
+
+    return dates, values
+
+def calc_f107a(f107_inst, f107_name='f107', f107a_name='f107a', min_pnts=41):
```

Comment:
👍

Reply:
Yeah, only problem is that I need to add in the provisional source of F10.7 now, because we have a big data gap!

Reply:
Added!